hydra-node/test/Hydra/ModelSpec.hs

```diff
@@ -0,0 +1,281 @@
+--        traceDump = printTrace (Proxy :: Proxy Tx) tr
+--    print traceDump
+--    True `shouldBe` True
+module Hydra.ModelSpec where
+
+import Hydra.Cardano.Api
+import Hydra.Prelude
+import Test.Hydra.Prelude hiding (after)
+
+-- This is completely safe
+import Unsafe.Coerce (unsafeCoerce)
+
+import qualified Cardano.Api.UTxO as UTxO
+import Control.Monad.IOSim (Failure (FailureException), IOSim, runSimTrace, traceResult)
+import Data.Map ((!))
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import Hydra.BehaviorSpec (TestHydraNode (..))
+import Hydra.Chain.Direct.Fixture (testNetworkId)
+import Hydra.ClientInput (ClientInput (..))
+import Hydra.Model (
+  Action (..),
+  GlobalState (..),
+  Nodes (Nodes, nodes),
+  OffChainState (..),
+  WorldState (..),
+  genPayment,
+ )
+import Hydra.Party (Party (..), deriveParty)
+import Hydra.ServerOutput (ServerOutput (..))
+import Test.QuickCheck (Property, counterexample, forAll, property, withMaxSuccess, within)
+import Test.QuickCheck.DynamicLogic (
+  DynFormula,
+  after,
+  afterAny,
+  done,
+  forAllQ,
+  forAllScripts,
+  withGenQ,
+ )
+import Test.QuickCheck.Gen.Unsafe (Capture (Capture), capture)
+import Test.QuickCheck.Monadic (PropertyM, assert, monadic, monadic', monitor, run)
+import Test.QuickCheck.StateModel (Actions, runActions, stateAfter, pattern Actions)
+import Test.Util (printTrace, traceInIOSim)
+import qualified Prelude
+
+spec :: Spec
+spec = do
+  prop "model generates consistent traces" $ withMaxSuccess 10000 prop_generateTraces
+  prop "implementation respects model" $ forAll arbitrary prop_checkModel
+  -- TODO: Implement this
+  prop "satisfies conflict-free liveness" $ property True
+
+prop_generateTraces :: AnyActions -> Property
+prop_generateTraces (AnyActions actions) =
+  let st = stateAfter actions
+   in case actions of
+        Actions [] -> property True
+        Actions _ ->
+          hydraState st /= Start
+            & counterexample ("state: " <> show st)
+
+prop_checkModel :: AnyActions -> Property
+prop_checkModel (AnyActions actions) =
+  within 2000000 $
+    property $
+      runIOSimProp $
+        monadic' $ do
+          (WorldState{hydraParties, hydraState}, _symEnv) <- runActions actions
+          run $ lift waitUntilTheEndOfTime
+          let parties = Set.fromList $ deriveParty . fst <$> hydraParties
+          nodes <- run $ gets nodes
+          assert (parties == Map.keysSet nodes)
+          forM_ parties $ \p -> do
+            assertNodeSeesAndReportsAllExpectedCommits hydraState nodes p
+            assertBalancesInOpenHeadAreConsistent hydraState nodes p
+
+prop_conflictFreeLiveness :: Property
+prop_conflictFreeLiveness =
+  forAllScripts (conflictFreeLiveness :: DynFormula (WorldState (IOSim s))) $ \actions -> do
+    monadic runPropertyM $ do
+      (_state, _env) <- runActions actions
+      pure True
+ where
+  runPropertyM :: StateT (Nodes m) m Property -> Property
+  runPropertyM = error "undefined"
```

Comment:
@abailly-iohk Why did you merge this already? The undefined here and the `property True` above indicate this is not ready for prime time yet.

Reply:
Was automatic because `qd-model` was fast-forward to this branch. I have reverted and re-established this version here as `dyn-logic` to be picked up again.